File: app/src/main/java/com/example/david/lists/view/common/ListViewBase.java
package com.example.david.lists.view.common;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.david.lists.R;
import com.example.david.lists.databinding.ListViewBaseBinding;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;

import javax.inject.Inject;
import javax.inject.Provider;

public abstract class ListViewBase extends Fragment
        implements TouchHelperCallback.MovementCallback {

    private ListViewBaseBinding binding;

    @Inject
    Provider<LinearLayoutManager> layoutManger;
    @Inject
    RecyclerView.ItemDecoration dividerItemDecorator;
    @Inject
    ItemTouchHelper itemTouchHelper;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.list_view_base, container, false);
        init();
        return binding.getRoot();
    }


    protected abstract void addButtonClicked();

    protected abstract void undoRecentDeletion();

    protected abstract void deletionNotificationTimedOut();

    protected abstract void draggingListItem(int fromPosition, int toPosition);

    protected abstract void permanentlyMoved(int newPosition);

    protected abstract String getTitle();

    protected abstract boolean enableUpNavigationOnToolbar();

    protected abstract RecyclerView.Adapter getAdapter();


    private void init() {
        initRecyclerView();
        initToolbar();
        initFab();
    }

    private void initRecyclerView() {
        RecyclerView recyclerView = binding.recyclerView;
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(layoutManger.get());
        recyclerView.addItemDecoration(dividerItemDecorator);
        itemTouchHelper.attachToRecyclerView(recyclerView);
        recyclerView.setAdapter(getAdapter());
    }

    private void initToolbar() {
        ((AppCompatActivity) getActivity()).setSupportActionBar(binding.toolbar);
        ActionBar actionBar = ((AppCompatActivity) getActivity()).getSupportActionBar();
        actionBar.setTitle(getTitle());
        actionBar.setDisplayHomeAsUpEnabled(enableUpNavigationOnToolbar());
    }

    private void initFab() {
        FloatingActionButton fab = binding.fab;
        fabClickListener(fab);
        fabScrollListener(fab);
    }

    private void fabClickListener(FloatingActionButton fab) {
        fab.setOnClickListener(view -> addButtonClicked());
    }

    private void fabScrollListener(FloatingActionButton fab) {
        binding.recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                if (dy > 0) {
                    fab.hide();
                } else if (dy < 0) {
                    fab.show();
                }
            }
        });
    }

    protected void notifyDeletionSnackbar(String message) {
        Snackbar.make(binding.rootLayout, message, Snackbar.LENGTH_LONG)
                .setAction(R.string.msg_undo, view -> undoRecentDeletion())
                .addCallback(new Snackbar.Callback() {
                    @Override
                    public void onDismissed(Snackbar transientBottomBar, int event) {
                        super.onDismissed(transientBottomBar, event);
                        if (validSnackbarEvent(event)) {
                            deletionNotificationTimedOut();
                        }
                    }
                })
                .show();
    }

    private boolean validSnackbarEvent(int event) {
        return event == Snackbar.Callback.DISMISS_EVENT_TIMEOUT
                || event == Snackbar.Callback.DISMISS_EVENT_SWIPE
                || event == Snackbar.Callback.DISMISS_EVENT_MANUAL;
    }


    @Override
    public void dragging(int fromPosition, int toPosition) {
        draggingListItem(fromPosition, toPosition);
    }

    @Override
    public void movedPermanently(int newPosition) {
        permanentlyMoved(newPosition);
    }


    protected void openDialogFragment(DialogFragment dialogFragment) {
        dialogFragment.setTargetFragment(this, 0);
        dialogFragment.show(getActivity().getSupportFragmentManager(), null);
    }


    protected void displayLoading() {
        binding.tvError.setVisibility(View.GONE);
        binding.recyclerView.setVisibility(View.GONE);
        binding.fab.hide();

        binding.progressBar.setVisibility(View.VISIBLE);
    }

    protected void displayList() {
        binding.progressBar.setVisibility(View.GONE);
        binding.tvError.setVisibility(View.GONE);

        binding.recyclerView.setVisibility(View.VISIBLE);
        binding.fab.show();
    }

    protected void displayError(String errorMessage) {
        binding.progressBar.setVisibility(View.GONE);
        binding.recyclerView.setVisibility(View.GONE);
        binding.fab.hide();
        
        binding.tvError.setText(errorMessage);
        binding.tvError.setVisibility(View.VISIBLE);
    }


    protected void toastMessage(String message) {
        Toast.makeText(getActivity().getApplicationContext(), message, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/example/david/lists/view/common/ListViewBase.kt
package com.example.david.lists.view.common

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.david.lists.R
import com.example.david.lists.common.toast
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.snackbar.Snackbar.Callback.*
import kotlinx.android.synthetic.main.list_view_base.*
import kotlinx.android.synthetic.main.toolbar.*
import javax.inject.Inject
import javax.inject.Provider

abstract class ListViewBase : Fragment(R.layout.list_view_base),
        TouchHelperCallback.MovementCallback {

    @Inject
    lateinit var layoutManger: Provider<LinearLayoutManager>

    @Inject
    lateinit var dividerItemDecorator: RecyclerView.ItemDecoration

    @Inject
    lateinit var itemTouchHelper: ItemTouchHelper

    protected abstract val title: String

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        init()
    }

    protected abstract fun addButtonClicked()

    protected abstract fun undoRecentDeletion()

    protected abstract fun deletionNotificationTimedOut()

    protected abstract fun draggingListItem(fromPosition: Int, toPosition: Int)

    protected abstract fun permanentlyMoved(newPosition: Int)

    protected abstract fun enableUpNavigationOnToolbar(): Boolean

    protected abstract fun getAdapter(): RecyclerView.Adapter<*>


    private fun init() {
        initRecyclerView()
        initToolbar()
        initFab()
    }

    private fun initRecyclerView() {
        recycler_view.apply {
            setHasFixedSize(true)
            layoutManager = layoutManger.get()
            addItemDecoration(dividerItemDecorator)
            itemTouchHelper.attachToRecyclerView(this)
            adapter = this@ListViewBase.getAdapter()
        }
    }

    private fun initToolbar() {
        with(toolbar) {
            (activity as AppCompatActivity).setSupportActionBar(this)
            title = this@ListViewBase.title
            if (enableUpNavigationOnToolbar()) {
                setNavigationIcon(R.drawable.ic_arrow_back_white_24dp)
                setNavigationOnClickListener { findNavController().navigateUp() }
            }
        }
    }

    private fun initFab() {
        fab.setOnClickListener { addButtonClicked() }
        fabScrollListener()
    }

    private fun fabScrollListener() {
        recycler_view.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                if (dy > 0) {
                    fab.hide()
                } else if (dy < 0) {
                    fab.show()
                }
            }
        })
    }

    protected fun notifyDeletionSnackbar(message: String) {
        Snackbar.make(root_layout, message, Snackbar.LENGTH_LONG)
                .setAction(R.string.msg_undo) { undoRecentDeletion() }
                .addCallback(object : Snackbar.Callback() {
                    override fun onDismissed(transientBottomBar: Snackbar?, event: Int) {
                        super.onDismissed(transientBottomBar, event)
                        if (validSnackbarEvent(event)) {
                            deletionNotificationTimedOut()
                        }
                    }
                })
                .show()
    }

    private fun validSnackbarEvent(event: Int) =
            event == DISMISS_EVENT_TIMEOUT
                    || event == DISMISS_EVENT_SWIPE
                    || event == DISMISS_EVENT_MANUAL


    override fun dragging(fromPosition: Int, toPosition: Int) {
        draggingListItem(fromPosition, toPosition)
    }

    override fun movedPermanently(newPosition: Int) {
        permanentlyMoved(newPosition)
    }


    /**
     * Because I am not using LiveData, the Logic classes will attempt to
     * manipulate the UI when the View is in an invalid state to do so.
     */
    private fun validLifecycleState() =
            lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)

    protected fun displayLoading() {
        if (validLifecycleState()) {
            tv_error.isGone = true
            recycler_view.isGone = true
            fab.hide()

            progress_bar.isVisible = true
        }
    }

    protected fun displayList() {
        if (validLifecycleState()) {
            progress_bar.isGone = true
            tv_error.isGone = true

            recycler_view.isVisible = true
            fab.show()
        }
    }

    protected fun displayError(errorMessage: String) {
        if (validLifecycleState()) {
            progress_bar?.isGone = true
            recycler_view?.isGone = true
            fab?.show()

            tv_error?.text = errorMessage
            tv_error?.isVisible = true
        }
    }


    protected fun toastMessage(message: String) {
        toast(message)
    }
}
